AnyText/AnyText.Core/Parser.DocumentSymbol.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection.Metadata;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using NMF.AnyText.Rules;
+using System.Diagnostics;
+using NMF.AnyText.Model;
+using System.Text.RegularExpressions;
+
+namespace NMF.AnyText
+{
+    public partial class Parser
+    {
+        private static Regex regex = new Regex(@"(\w+) '(\w+)'");
+
+        public DocumentSymbol[] GetDocumentSymbolsFromRoot()
+        {
+            RuleApplication rootApplication = Context.RootRuleApplication;
+
+            if (rootApplication.IsPositive)
+            {
+                var parsedDocumentSymbols = GetDocumentSymbols(rootApplication);
+                return parsedDocumentSymbols;
+            }
+
+            return Array.Empty<DocumentSymbol>();
+        }
+
+        private DocumentSymbol[] GetDocumentSymbols(RuleApplication ruleApplication)
+        {
+            if (ruleApplication is MultiRuleApplication multiRuleApplication)
+            {
+                if (ruleApplication.Rule is OneOrMoreRule)
+                {
+                    return multiRuleApplication.Inner.Select(innerRuleApplication => GetDocumentSymbols(innerRuleApplication)).SelectMany(i => i).ToArray();
+                }
+
+                if (ruleApplication.Rule.SymbolKind == SymbolKind.Null)
+                {
+                    return Array.Empty<DocumentSymbol>();
+                }
+
+                var literal = ruleApplication.GetValue(Context).ToString();
```

Comment:
Damit hängen wir sehr nah an der ToString-Implementierung von Modellen, die AnyText.Core aber eigentlich gar nicht kennt. Können wir das nicht so machen, dass wir die ganzen RuleApplications durchgehen und die jeweiligen Rules immer fragen, ob sie vielleicht für die RuleApplication ein Symbol eintragen möchten? Falls ja, dann könnte die generische Methode den Range und die SelectionRange ja übertragen, so wie das hier jetzt schon implementiert ist.

Am liebsten wäre mir, wenn die Rule einfach ein DocumentSymbol oder null zurückgeben würde und eben nur die ModelElementRule tatsächlich ein Symbol zurückliefert, alle anderen null. Falls es nicht null ist, kommt es in die Liste.